# Mutual analysis of interaction networks and quantitative trait loci for budding yeast

---
## Table of contents
* [Description](#description)
* [Motivation](#motivation)
* [How to use this?](#how-to-use-this)

---
## Description
![Saccharomyces cerevisiae, the budding yeast](../img/qtl_distributions/eQTLs_2011_vs_2018.png)

This is my first project in systems biology. I started it during the second year of my undergraduate studies in Moscow 
Institute of Physics and Technology ([MIPT](https://mipt.ru/english/), in Russia better known as "Физтех"). It's a work-in-progress, because workload at our university is really intense. It will eventually become my bachelor's thesis and will be thoroughly documented and tested by that time. It's my first experience in maintaining big, long-term projects, thereby I still have much to learn, but gradually I will adopt and start following software engineering best practices. Moreover, as it usually happens in scientific programming, the code is often being rewritten or thrown away. As a result, most of the work is done behind the curtain in the form of weekly reports I use to communicate new results and ideas to by supervisor. 

This research is carried out under supervision of [Yuri Pritykin](https://scholar.google.com/citations?hl=en&user=Arx56RkJBrYC&view_op=list_works&sortby=pubdate), PhD (Princeton), Research Scholar at MSKCC.

This repository is maintained to ensure the reproducibility of our results.
Project's main language is Python. We use Jupyter Notebook for exploratory data analysis. We utilize MIPT supercomputing capabilities to carry out computational experiments. Implementation of most of the of the statistical tests is efficiently parallelized to fully expolit their embrassingly parallel nature. 


[↥ back to top](#table-of-contents)

---
## Motivation

* **Why QTLs?** Quantitative trait loci or QTLs are a hot topic in contemporary biological research. The link between 
genotype and phenotype, they attracted increasing attention in recent years due to developments of high-throughouput 
technologies like RNA-seq. This research deals with eQTLs — with mRNA expression as quantitative trait — and pQTLs — 
same for protein expression. Genetic architectures of eQTLs and pQTLs are very different, despite the obvious 
relationship between transcription and translation of the genes. There is an ongoing debate on this matter in contemporary research.
We decided to make a contribution by comparing genetic architectures underlying eQTLs and pQTLs linked to functional
modules — sets of interacting genes participating in a particular cell process. Our approach is the possible solution 
that reduces eQTL/pQTL discrepancy via protein-protein interaction networks.          
* **Why yeast?** Despite our hypothesis isn't specific to this species, we concentrated on *Saccharomyces cerevisiae* 
or *baker's yeast*. It's a popular model organism, sequenced to unprecedented depth. Most of its genes are 
annotated and their interactions are thoroughly studied. Yeast biological databases are well-developed, carefully 
curated and easy to work with, which results in increasing quality of computational experiments.  
* **Why does this matter?** Some complex diseases can be tracked down to genetic level — either mRNA or protein levels
of particular genes are wrong. Understanding the relationships between these two is necessary for the accurate treatment. 
eQTLs have been studied for the long time already, experiment designs are standardized and well-known. 
pQTLs are a different matter, because protein expression measurements are still costly and carry a lot
of noise. Introduction of additional structure — the interactome — helps us to make the most out of limited data. 
It may be, that to bridge the gap between eQTLs and pQTLs we should look at them on a bigger scale.             
* **Why us?** We have an unusually rich protein expression dataset. It drastically increases the statistical 
power of our observations. We will also try using the latest algorithms for QTL mapping to improve precision of our results.

[↥ back to top](#table-of-contents)

---
## Project Structure

```text
|__ data 
        |__ raw — files from papers or biological databases with almost no preprocessing
        |__ tmp — temporary files (joblib caches, intermediate results of batch processing etc.)
        |__ processed — autogenerated files; editing those manually is forbidden
|__ img — autogenerated images; some of those are used in the reports
|__ lib — custom libraries used throughout the project
|__ notebooks — jupyter notebooks where most of the analysis is done
|__ reports — Google-Docs with weekly progress reports 
|__ results
        |__ raw — autogenerated files; editing those manually is forbidden
        |__ processed — files that can be included in reports
|__ utils — utilities, mainly for SLURM-related tasks
|__ src — R and Python scripts
``` 

[↥ back to top](#table-of-contents)

---
