{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this notebook the eQTL identification is performed.\n",
    "\n",
    "General workflow description: for each marker identify whether it is a QTL for some expressed gene by running a statistical test over sets of expression values of each of them, divided by inheritance pattern in a strain, and applying FDR-correction thereafter "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import cm\n",
    "import numpy as np\n",
    "import time\n",
    "from scipy import stats\n",
    "import networkx as nx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# BY and RM strains values must be averaged and placed into two separate columns\n",
    "\n",
    "expression_df = pd.read_table('./data/brem2005_RNA_expression.csv')\n",
    "genotypes_df = pd.read_table('./data/jbloom_strains_genotyped.tab')\n",
    "\n",
    "markers_n = genotypes_df.shape[0]\n",
    "rna_n = expression_df.shape[0]\n",
    "\n",
    "# Add columns with the parental genotypes\n",
    "# to use their averaged expression data\n",
    "\n",
    "genotypes_df[\"BY\"] = pd.DataFrame(data=np.zeros(markers_n, dtype=\"int\")) \n",
    "genotypes_df[\"RM\"] = pd.DataFrame(data=np.ones(markers_n, dtype=\"int\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_loc = dict(zip(genotypes_df[\"RQTL_name\"], np.arange(markers_n)))\n",
    "RNA_loc = dict(zip(expression_df[\"IDENTIFIER\"], np.arange(rna_n)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Average the expression values over all parental strains.\n",
    "# Add averaged columns called \"BY\" and \"RM\" to expression dataframe\n",
    "# and remove the parental strains' columns\n",
    "\n",
    "BY_expr_df = expression_df.iloc[:, 2:8]\n",
    "RM_expr_df = expression_df.iloc[:, 8:20]\n",
    "\n",
    "\n",
    "def avg_expr_df(df):\n",
    "    row_n, col_n = df.shape \n",
    "    averaged = np.full(shape=row_n, fill_value=np.nan)\n",
    "    # iterate over rows of a dataframe\n",
    "    for i, row in enumerate(df.itertuples()):\n",
    "        expr_sum, cnt = 0., 0.\n",
    "        # and average non-NaN values of the columns\n",
    "        # presence of NaNs is exactly the reason why the columns \n",
    "        # cannot be simply added together and averaged\n",
    "        for j in range(1, col_n + 1):\n",
    "            if not np.isnan(row[j]):\n",
    "                expr_sum += row[j]\n",
    "                cnt += 1\n",
    "        if cnt:\n",
    "            averaged[i] = expr_sum / cnt\n",
    "    return averaged\n",
    "\n",
    "\n",
    "expression_df[\"BY\"] = pd.DataFrame(avg_expr_df(BY_expr_df))\n",
    "expression_df[\"RM\"] = pd.DataFrame(avg_expr_df(RM_expr_df))\n",
    "\n",
    "# Non-averaged expression data of parental strains is no longer necessary,\n",
    "# so the appropriate columns are dropped from the dataframe\n",
    "expression_df = expression_df.drop(\n",
    "    BY_expr_df.columns.tolist() + RM_expr_df.columns.tolist(), \n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Haven't I cleaned the data before exactly not to write these lines?\n",
    "\n",
    "# strain_data = pd.read_csv('./data/brem2005_strains.csv')\n",
    "# strain_names = strain_data['Strain name'].tolist()\n",
    "# progeny_names = strain_names[18:]\n",
    "\n",
    "strain_names = expression_df.columns.tolist()[2:]\n",
    "progeny_names = strain_names[18:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_REF</th>\n",
       "      <th>IDENTIFIER</th>\n",
       "      <th>1_1_d</th>\n",
       "      <th>1_3_d</th>\n",
       "      <th>1_4_d</th>\n",
       "      <th>1_5_c</th>\n",
       "      <th>2_2_d</th>\n",
       "      <th>2_3_d</th>\n",
       "      <th>2_4_a</th>\n",
       "      <th>2_5_d</th>\n",
       "      <th>...</th>\n",
       "      <th>21_4_d</th>\n",
       "      <th>21_5_c</th>\n",
       "      <th>22_2_d</th>\n",
       "      <th>22_3_b</th>\n",
       "      <th>22_4_d</th>\n",
       "      <th>22_5_d</th>\n",
       "      <th>23_3_d</th>\n",
       "      <th>23_5_d</th>\n",
       "      <th>24_1_d</th>\n",
       "      <th>25_1_d</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TRS120</td>\n",
       "      <td>0.1235</td>\n",
       "      <td>0.0790</td>\n",
       "      <td>-0.0230</td>\n",
       "      <td>0.1050</td>\n",
       "      <td>0.1205</td>\n",
       "      <td>0.1425</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>-0.2760</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.0330</td>\n",
       "      <td>0.2555</td>\n",
       "      <td>0.3165</td>\n",
       "      <td>0.1775</td>\n",
       "      <td>0.0615</td>\n",
       "      <td>0.0640</td>\n",
       "      <td>-0.4475</td>\n",
       "      <td>0.109</td>\n",
       "      <td>0.1445</td>\n",
       "      <td>0.3130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>SCC2</td>\n",
       "      <td>0.2955</td>\n",
       "      <td>0.3965</td>\n",
       "      <td>-0.0795</td>\n",
       "      <td>0.0530</td>\n",
       "      <td>0.3795</td>\n",
       "      <td>0.2050</td>\n",
       "      <td>0.1655</td>\n",
       "      <td>-0.2225</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2935</td>\n",
       "      <td>0.5400</td>\n",
       "      <td>0.4230</td>\n",
       "      <td>0.2575</td>\n",
       "      <td>0.0770</td>\n",
       "      <td>0.2280</td>\n",
       "      <td>0.3815</td>\n",
       "      <td>0.239</td>\n",
       "      <td>0.1045</td>\n",
       "      <td>0.3535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>FLO1</td>\n",
       "      <td>0.7180</td>\n",
       "      <td>-0.0090</td>\n",
       "      <td>0.9030</td>\n",
       "      <td>0.1855</td>\n",
       "      <td>0.9040</td>\n",
       "      <td>0.2455</td>\n",
       "      <td>-1.6495</td>\n",
       "      <td>-0.0360</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.1550</td>\n",
       "      <td>-0.1370</td>\n",
       "      <td>0.4565</td>\n",
       "      <td>0.5520</td>\n",
       "      <td>0.0735</td>\n",
       "      <td>0.2610</td>\n",
       "      <td>-0.0545</td>\n",
       "      <td>0.074</td>\n",
       "      <td>0.2690</td>\n",
       "      <td>0.4370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>MYO3</td>\n",
       "      <td>-0.0220</td>\n",
       "      <td>0.2745</td>\n",
       "      <td>0.3285</td>\n",
       "      <td>0.3425</td>\n",
       "      <td>0.1615</td>\n",
       "      <td>0.3145</td>\n",
       "      <td>-0.1400</td>\n",
       "      <td>0.1160</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.0880</td>\n",
       "      <td>-0.1340</td>\n",
       "      <td>0.0800</td>\n",
       "      <td>0.0725</td>\n",
       "      <td>-0.0875</td>\n",
       "      <td>-0.0995</td>\n",
       "      <td>-0.5545</td>\n",
       "      <td>0.132</td>\n",
       "      <td>0.2220</td>\n",
       "      <td>0.2430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>PDR10</td>\n",
       "      <td>-0.0130</td>\n",
       "      <td>0.2100</td>\n",
       "      <td>0.2540</td>\n",
       "      <td>0.2925</td>\n",
       "      <td>0.6040</td>\n",
       "      <td>0.4395</td>\n",
       "      <td>-0.1370</td>\n",
       "      <td>-0.0465</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.0610</td>\n",
       "      <td>-0.1505</td>\n",
       "      <td>-0.2370</td>\n",
       "      <td>-0.0255</td>\n",
       "      <td>0.0360</td>\n",
       "      <td>0.1130</td>\n",
       "      <td>0.1615</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.2985</td>\n",
       "      <td>-0.0760</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 110 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_REF</th>\n",
       "      <th>IDENTIFIER</th>\n",
       "      <th>1_1_d</th>\n",
       "      <th>1_3_d</th>\n",
       "      <th>1_4_d</th>\n",
       "      <th>1_5_c</th>\n",
       "      <th>2_2_d</th>\n",
       "      <th>2_3_d</th>\n",
       "      <th>2_4_a</th>\n",
       "      <th>2_5_d</th>\n",
       "      <th>...</th>\n",
       "      <th>21_4_d</th>\n",
       "      <th>21_5_c</th>\n",
       "      <th>22_2_d</th>\n",
       "      <th>22_3_b</th>\n",
       "      <th>22_4_d</th>\n",
       "      <th>22_5_d</th>\n",
       "      <th>23_3_d</th>\n",
       "      <th>23_5_d</th>\n",
       "      <th>24_1_d</th>\n",
       "      <th>25_1_d</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TRS120</td>\n",
       "      <td>0.1235</td>\n",
       "      <td>0.0790</td>\n",
       "      <td>-0.0230</td>\n",
       "      <td>0.1050</td>\n",
       "      <td>0.1205</td>\n",
       "      <td>0.1425</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>-0.2760</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.0330</td>\n",
       "      <td>0.2555</td>\n",
       "      <td>0.3165</td>\n",
       "      <td>0.1775</td>\n",
       "      <td>0.0615</td>\n",
       "      <td>0.0640</td>\n",
       "      <td>-0.4475</td>\n",
       "      <td>0.109</td>\n",
       "      <td>0.1445</td>\n",
       "      <td>0.3130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>SCC2</td>\n",
       "      <td>0.2955</td>\n",
       "      <td>0.3965</td>\n",
       "      <td>-0.0795</td>\n",
       "      <td>0.0530</td>\n",
       "      <td>0.3795</td>\n",
       "      <td>0.2050</td>\n",
       "      <td>0.1655</td>\n",
       "      <td>-0.2225</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2935</td>\n",
       "      <td>0.5400</td>\n",
       "      <td>0.4230</td>\n",
       "      <td>0.2575</td>\n",
       "      <td>0.0770</td>\n",
       "      <td>0.2280</td>\n",
       "      <td>0.3815</td>\n",
       "      <td>0.239</td>\n",
       "      <td>0.1045</td>\n",
       "      <td>0.3535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>FLO1</td>\n",
       "      <td>0.7180</td>\n",
       "      <td>-0.0090</td>\n",
       "      <td>0.9030</td>\n",
       "      <td>0.1855</td>\n",
       "      <td>0.9040</td>\n",
       "      <td>0.2455</td>\n",
       "      <td>-1.6495</td>\n",
       "      <td>-0.0360</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.1550</td>\n",
       "      <td>-0.1370</td>\n",
       "      <td>0.4565</td>\n",
       "      <td>0.5520</td>\n",
       "      <td>0.0735</td>\n",
       "      <td>0.2610</td>\n",
       "      <td>-0.0545</td>\n",
       "      <td>0.074</td>\n",
       "      <td>0.2690</td>\n",
       "      <td>0.4370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>MYO3</td>\n",
       "      <td>-0.0220</td>\n",
       "      <td>0.2745</td>\n",
       "      <td>0.3285</td>\n",
       "      <td>0.3425</td>\n",
       "      <td>0.1615</td>\n",
       "      <td>0.3145</td>\n",
       "      <td>-0.1400</td>\n",
       "      <td>0.1160</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.0880</td>\n",
       "      <td>-0.1340</td>\n",
       "      <td>0.0800</td>\n",
       "      <td>0.0725</td>\n",
       "      <td>-0.0875</td>\n",
       "      <td>-0.0995</td>\n",
       "      <td>-0.5545</td>\n",
       "      <td>0.132</td>\n",
       "      <td>0.2220</td>\n",
       "      <td>0.2430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>PDR10</td>\n",
       "      <td>-0.0130</td>\n",
       "      <td>0.2100</td>\n",
       "      <td>0.2540</td>\n",
       "      <td>0.2925</td>\n",
       "      <td>0.6040</td>\n",
       "      <td>0.4395</td>\n",
       "      <td>-0.1370</td>\n",
       "      <td>-0.0465</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.0610</td>\n",
       "      <td>-0.1505</td>\n",
       "      <td>-0.2370</td>\n",
       "      <td>-0.0255</td>\n",
       "      <td>0.0360</td>\n",
       "      <td>0.1130</td>\n",
       "      <td>0.1615</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.2985</td>\n",
       "      <td>-0.0760</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 110 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_df.iloc[:, 0:110].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RQTL_name</th>\n",
       "      <th>2_7_a</th>\n",
       "      <th>2_7_b</th>\n",
       "      <th>2_7_c</th>\n",
       "      <th>2_7_d</th>\n",
       "      <th>21_5_c</th>\n",
       "      <th>22_1_d</th>\n",
       "      <th>22_2_d</th>\n",
       "      <th>19_2_c</th>\n",
       "      <th>19_3_c</th>\n",
       "      <th>...</th>\n",
       "      <th>9_7_d</th>\n",
       "      <th>26_2_d</th>\n",
       "      <th>19_1_c</th>\n",
       "      <th>21_3_d</th>\n",
       "      <th>21_1_d</th>\n",
       "      <th>15_6_c</th>\n",
       "      <th>17_1_a</th>\n",
       "      <th>name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YAL069W_1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>1</td>\n",
       "      <td>483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YAL069W_2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>1</td>\n",
       "      <td>484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NAL013C_3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NAL013C</td>\n",
       "      <td>1</td>\n",
       "      <td>3220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NAL013C_4</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NAL013C</td>\n",
       "      <td>1</td>\n",
       "      <td>3223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NAL013C_5</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NAL013C</td>\n",
       "      <td>1</td>\n",
       "      <td>3232</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 128 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RQTL_name</th>\n",
       "      <th>2_7_a</th>\n",
       "      <th>2_7_b</th>\n",
       "      <th>2_7_c</th>\n",
       "      <th>2_7_d</th>\n",
       "      <th>21_5_c</th>\n",
       "      <th>22_1_d</th>\n",
       "      <th>22_2_d</th>\n",
       "      <th>19_2_c</th>\n",
       "      <th>19_3_c</th>\n",
       "      <th>...</th>\n",
       "      <th>9_7_d</th>\n",
       "      <th>26_2_d</th>\n",
       "      <th>19_1_c</th>\n",
       "      <th>21_3_d</th>\n",
       "      <th>21_1_d</th>\n",
       "      <th>15_6_c</th>\n",
       "      <th>17_1_a</th>\n",
       "      <th>name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YAL069W_1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>1</td>\n",
       "      <td>483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YAL069W_2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>1</td>\n",
       "      <td>484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NAL013C_3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NAL013C</td>\n",
       "      <td>1</td>\n",
       "      <td>3220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NAL013C_4</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NAL013C</td>\n",
       "      <td>1</td>\n",
       "      <td>3223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NAL013C_5</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NAL013C</td>\n",
       "      <td>1</td>\n",
       "      <td>3232</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 128 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_df.iloc[:, 0:128].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Divide all progeny into groups by their inheritance pattern\n",
    "# for a given genetic marker, and then plot the data clouds\n",
    "# to visually observe if there is any correlation between marker\n",
    "# and RNA expression \n",
    "\n",
    "# What is the asymptotics of this function?\n",
    "\n",
    "# Divide expression data for a given gene in two groups,\n",
    "# based on inheritance pattern of a given marker\n",
    "def expression_by_RNA_and_marker(RNA_name, marker_name):\n",
    "    from_BY, from_RM = [], []\n",
    "    RNA_pos = expression_df[expression_df.IDENTIFIER == RNA_name].index.tolist()[0]\n",
    "    marker_pos = genotypes_df[genotypes_df.RQTL_name == marker_name].index.tolist()[0]\n",
    "    for strain in progeny_names:\n",
    "        # Optimization: extract the corresponding rows in advance\n",
    "        expression_value = expression_df[strain].iloc[RNA_pos]\n",
    "        inheritance_pattern = genotypes_df[strain].iloc[marker_pos]\n",
    "        if np.isnan(expression_value):\n",
    "            continue\n",
    "        if inheritance_pattern == 0:\n",
    "            from_BY.append(expression_value)\n",
    "        elif inheritance_pattern == 1:\n",
    "            from_RM.append(expression_value)\n",
    "    return from_BY, from_RM\n",
    "\n",
    "\n",
    "# For the given pair (expressed gene, marker) test the hypothesis\n",
    "# that inherited variant of a marker influences gene expression significantly \n",
    "def test_linkage(RNA_name, marker_name, eps=1e-5):\n",
    "    from_BY, from_RM = expression_by_RNA_and_marker(RNA_name, marker_name)\n",
    "    statistic, pvalue = stats.mannwhitneyu(x=from_BY, y=from_RM)\n",
    "    return (pvalue <= eps, pvalue)\n",
    "\n",
    "\n",
    "# Divide expression data by inherited marker \n",
    "# variant and then plot the resulting groups \n",
    "def plot_expression_to_marker_correlation(RNA_name, marker_name):\n",
    "    from_BY, from_RM = expression_by_RNA_and_marker(RNA_name, marker_name)\n",
    "    xlabels = np.append(\n",
    "                    np.full((1, len(from_BY)), 1), \n",
    "                    np.full((1, len(from_RM)), 2))\\\n",
    "                    + np.random.normal(0, 0.01, len(from_BY) + len(from_RM)\n",
    "            )  \n",
    "    ylabels = np.array(from_BY + from_RM)\n",
    "    plt.figure(figsize=(20, 10))\n",
    "    plt.rcParams[\"axes.facecolor\"] = 'white'\n",
    "    plt.title(\"p-value: {}\".format(pvalue))\n",
    "    plt.xlabel(\"class label\")\n",
    "    plt.ylabel(\"expression value\")\n",
    "    plt.scatter(\n",
    "        x=xlabels, y=ylabels,\n",
    "        c=ylabels, cmap=cm.jet\n",
    "    )\n",
    "    plt.savefig(\"./img/\" + RNA_name + \"_to_\" + marker_name + \".png\")\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**TODO**:\n",
    "    1.  Find QTLs for every gene using 5% pvalue threshold\n",
    "    2.  Validate QTLs using permutations test and calculate FDR\n",
    "    3.  Construct and visualize a bipartite graph visualizing \n",
    "        the linkages found during experiments \n",
    "    4.  Also, it's worth calculating the linkages for every\n",
    "        marker gene and plotting the bar chart, placing markers\n",
    "        accordingly to their position in the chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# When and how to apply multiple-testing correction (FDR, q-value)?\n",
    "\n",
    "# Set the seed to ensure either reproducibility\n",
    "# or randomness of the generated sample \n",
    "np.random.seed(int(time.time()))\n",
    "\n",
    "# Extract subsample\n",
    "test_pairs = list(\n",
    "                zip( \n",
    "                    np.random.choice(expression_df[\"IDENTIFIER\"].tolist(), 2000, replace=True),\n",
    "                    np.random.choice(genotypes_df[\"RQTL_name\"].tolist(), 2000, replace=True)\n",
    "                )\n",
    "            )\n",
    "\n",
    "# Check linkage for all entries of this subsample\n",
    "# and construct a bipartite graph of interactions\n",
    "linkage_graph = nx.Graph()\n",
    "for RNA_name, marker_name in test_pairs:\n",
    "    result, pvalue = test_linkage(RNA_name, marker_name, eps=0.05)\n",
    "    if result:\n",
    "        if not linkage_graph.has_node(RNA_name):\n",
    "            linkage_graph.add_node(RNA_name, bipartite=0)\n",
    "        if not linkage_graph.has_node(marker_name):\n",
    "            linkage_graph.add_node(marker_name, bipartite=1)\n",
    "        linkage_graph.add_edge(RNA_name, marker_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/vvi/anaconda3/lib/python3.5/site-packages/networkx/drawing/nx_pylab.py:126: MatplotlibDeprecationWarning: pyplot.hold is deprecated.\n    Future behavior will be consistent with the long-time default:\n    plot commands add elements without first clearing the\n    Axes and/or Figure.\n  b = plt.ishold()\n/home/vvi/anaconda3/lib/python3.5/site-packages/networkx/drawing/nx_pylab.py:138: MatplotlibDeprecationWarning: pyplot.hold is deprecated.\n    Future behavior will be consistent with the long-time default:\n    plot commands add elements without first clearing the\n    Axes and/or Figure.\n  plt.hold(b)\n/home/vvi/anaconda3/lib/python3.5/site-packages/matplotlib/__init__.py:917: UserWarning: axes.hold is deprecated. Please remove it from your matplotlibrc and/or style files.\n  warnings.warn(self.msg_depr_set % key)\n/home/vvi/anaconda3/lib/python3.5/site-packages/matplotlib/rcsetup.py:152: UserWarning: axes.hold is deprecated, will be removed in 3.0\n  warnings.warn(\"axes.hold is deprecated, will be removed in 3.0\")\n"
     ]
    }
   ],
   "source": [
    "# Built-in bipartite.sets() works strangely \n",
    "# maybe, it's only so for undirected graphs,\n",
    "# I should check that on some toy example\n",
    "\n",
    "top_v, bottom_v = [], []\n",
    "for node, data in linkage_graph.nodes(data=True):\n",
    "    if data[\"bipartite\"] == 0:\n",
    "        bottom_v.append(node)\n",
    "    else:\n",
    "        top_v.append(node)\n",
    "        \n",
    "# To plot a bipartite graph correctly, the positions\n",
    "# of the vertices must be written down explicitly\n",
    "\n",
    "pos = dict()\n",
    "pos.update((n, (1, 2*i)) for i, n in enumerate(top_v))\n",
    "pos.update((n, (2, 2*i)) for i, n in enumerate(bottom_v))\n",
    "\n",
    "# Maybe, (20, 20) is too small: the labels overlap significantly\n",
    "\n",
    "plt.figure(figsize=(20, 30))\n",
    "nx.draw(\n",
    "    linkage_graph,\n",
    "    with_labels=True,\n",
    "    node_size=50,\n",
    "    edge_width=3.0,\n",
    "    pos=pos,\n",
    "    node_color=list(linkage_graph.degree().values()),\n",
    "    edge_color='b',\n",
    "    cmap=plt.cm.Blues,\n",
    "    alpha=0.5\n",
    ")\n",
    "plt.savefig(\"./img/graph.png\")\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Extract the marker-nodes and number of linkages to them\n",
    "# preserving their order based on genome location\n",
    "\n",
    "marker_nodes = sorted(\n",
    "    list(linkage_graph.degree(top_v).items()), \n",
    "    key=lambda p: marker_loc[p[0]]\n",
    ")\n",
    "\n",
    "# Pythonic way of unzipping a list of tuples\n",
    "# into two separate lists of their coordinates\n",
    "\n",
    "m_names, m_degrees = map(list, zip(*marker_nodes))  \n",
    "\n",
    "plt.figure(figsize=(40, 20))\n",
    "plt.plot(m_degrees)\n",
    "plt.xticks(\n",
    "    range(len(marker_nodes)), \n",
    "    m_names,\n",
    "    rotation=\"vertical\"\n",
    ")\n",
    "plt.savefig(\"./img/linkage_map.png\")\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
