{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import cm\n",
    "import numpy as np\n",
    "import time\n",
    "from scipy import stats\n",
    "import networkx as nx\n",
    "from networkx.algorithms import bipartite\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\", category=DeprecationWarning) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_df = pd.read_table('./data/brem2005_RNA_expression.csv')\n",
    "genotypes_df = pd.read_table('./data/jbloom_strains_genotyped.tab')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_data = pd.read_csv('./data/brem2005_strains.csv')\n",
    "strain_names = strain_data['Strain name'].tolist()\n",
    "progeny_names = strain_names[18:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_REF</th>\n",
       "      <th>IDENTIFIER</th>\n",
       "      <th>By1</th>\n",
       "      <th>By2</th>\n",
       "      <th>BY3</th>\n",
       "      <th>BY4</th>\n",
       "      <th>BY5</th>\n",
       "      <th>BY6</th>\n",
       "      <th>RM1_1</th>\n",
       "      <th>RM2_1</th>\n",
       "      <th>...</th>\n",
       "      <th>22_4_d</th>\n",
       "      <th>22_5_d</th>\n",
       "      <th>23_3_d</th>\n",
       "      <th>23_5_d</th>\n",
       "      <th>24_1_d</th>\n",
       "      <th>25_1_d</th>\n",
       "      <th>25_3_d</th>\n",
       "      <th>25_4_d</th>\n",
       "      <th>26_1_d</th>\n",
       "      <th>26_2_d</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TRS120</td>\n",
       "      <td>-0.0620</td>\n",
       "      <td>-0.094</td>\n",
       "      <td>0.0925</td>\n",
       "      <td>-0.1135</td>\n",
       "      <td>-0.289</td>\n",
       "      <td>-0.0420</td>\n",
       "      <td>-0.0570</td>\n",
       "      <td>-0.0140</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0615</td>\n",
       "      <td>0.0640</td>\n",
       "      <td>-0.4475</td>\n",
       "      <td>0.109</td>\n",
       "      <td>0.1445</td>\n",
       "      <td>0.3130</td>\n",
       "      <td>-0.0185</td>\n",
       "      <td>-0.0360</td>\n",
       "      <td>0.1210</td>\n",
       "      <td>0.0575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>SCC2</td>\n",
       "      <td>0.1340</td>\n",
       "      <td>0.066</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.0460</td>\n",
       "      <td>-0.409</td>\n",
       "      <td>-0.4100</td>\n",
       "      <td>0.4235</td>\n",
       "      <td>0.2045</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0770</td>\n",
       "      <td>0.2280</td>\n",
       "      <td>0.3815</td>\n",
       "      <td>0.239</td>\n",
       "      <td>0.1045</td>\n",
       "      <td>0.3535</td>\n",
       "      <td>-0.0300</td>\n",
       "      <td>0.0775</td>\n",
       "      <td>0.4845</td>\n",
       "      <td>0.1150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>FLO1</td>\n",
       "      <td>-0.1165</td>\n",
       "      <td>-0.209</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.0050</td>\n",
       "      <td>0.078</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.8970</td>\n",
       "      <td>0.6460</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0735</td>\n",
       "      <td>0.2610</td>\n",
       "      <td>-0.0545</td>\n",
       "      <td>0.074</td>\n",
       "      <td>0.2690</td>\n",
       "      <td>0.4370</td>\n",
       "      <td>0.2475</td>\n",
       "      <td>0.0840</td>\n",
       "      <td>-0.1570</td>\n",
       "      <td>0.0040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>MYO3</td>\n",
       "      <td>-0.1595</td>\n",
       "      <td>-0.228</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1830</td>\n",
       "      <td>-0.102</td>\n",
       "      <td>0.1815</td>\n",
       "      <td>-0.1555</td>\n",
       "      <td>0.0445</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.0875</td>\n",
       "      <td>-0.0995</td>\n",
       "      <td>-0.5545</td>\n",
       "      <td>0.132</td>\n",
       "      <td>0.2220</td>\n",
       "      <td>0.2430</td>\n",
       "      <td>0.1830</td>\n",
       "      <td>-0.0040</td>\n",
       "      <td>-0.1085</td>\n",
       "      <td>0.2220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>PDR10</td>\n",
       "      <td>0.0395</td>\n",
       "      <td>-0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0430</td>\n",
       "      <td>-0.162</td>\n",
       "      <td>0.1560</td>\n",
       "      <td>0.2075</td>\n",
       "      <td>0.1125</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0360</td>\n",
       "      <td>0.1130</td>\n",
       "      <td>0.1615</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.2985</td>\n",
       "      <td>-0.0760</td>\n",
       "      <td>-0.2335</td>\n",
       "      <td>0.1015</td>\n",
       "      <td>-0.2010</td>\n",
       "      <td>-0.3375</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 132 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_REF</th>\n",
       "      <th>IDENTIFIER</th>\n",
       "      <th>By1</th>\n",
       "      <th>By2</th>\n",
       "      <th>BY3</th>\n",
       "      <th>BY4</th>\n",
       "      <th>BY5</th>\n",
       "      <th>BY6</th>\n",
       "      <th>RM1_1</th>\n",
       "      <th>RM2_1</th>\n",
       "      <th>...</th>\n",
       "      <th>22_4_d</th>\n",
       "      <th>22_5_d</th>\n",
       "      <th>23_3_d</th>\n",
       "      <th>23_5_d</th>\n",
       "      <th>24_1_d</th>\n",
       "      <th>25_1_d</th>\n",
       "      <th>25_3_d</th>\n",
       "      <th>25_4_d</th>\n",
       "      <th>26_1_d</th>\n",
       "      <th>26_2_d</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TRS120</td>\n",
       "      <td>-0.0620</td>\n",
       "      <td>-0.094</td>\n",
       "      <td>0.0925</td>\n",
       "      <td>-0.1135</td>\n",
       "      <td>-0.289</td>\n",
       "      <td>-0.0420</td>\n",
       "      <td>-0.0570</td>\n",
       "      <td>-0.0140</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0615</td>\n",
       "      <td>0.0640</td>\n",
       "      <td>-0.4475</td>\n",
       "      <td>0.109</td>\n",
       "      <td>0.1445</td>\n",
       "      <td>0.3130</td>\n",
       "      <td>-0.0185</td>\n",
       "      <td>-0.0360</td>\n",
       "      <td>0.1210</td>\n",
       "      <td>0.0575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>SCC2</td>\n",
       "      <td>0.1340</td>\n",
       "      <td>0.066</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.0460</td>\n",
       "      <td>-0.409</td>\n",
       "      <td>-0.4100</td>\n",
       "      <td>0.4235</td>\n",
       "      <td>0.2045</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0770</td>\n",
       "      <td>0.2280</td>\n",
       "      <td>0.3815</td>\n",
       "      <td>0.239</td>\n",
       "      <td>0.1045</td>\n",
       "      <td>0.3535</td>\n",
       "      <td>-0.0300</td>\n",
       "      <td>0.0775</td>\n",
       "      <td>0.4845</td>\n",
       "      <td>0.1150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>FLO1</td>\n",
       "      <td>-0.1165</td>\n",
       "      <td>-0.209</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.0050</td>\n",
       "      <td>0.078</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.8970</td>\n",
       "      <td>0.6460</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0735</td>\n",
       "      <td>0.2610</td>\n",
       "      <td>-0.0545</td>\n",
       "      <td>0.074</td>\n",
       "      <td>0.2690</td>\n",
       "      <td>0.4370</td>\n",
       "      <td>0.2475</td>\n",
       "      <td>0.0840</td>\n",
       "      <td>-0.1570</td>\n",
       "      <td>0.0040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>MYO3</td>\n",
       "      <td>-0.1595</td>\n",
       "      <td>-0.228</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1830</td>\n",
       "      <td>-0.102</td>\n",
       "      <td>0.1815</td>\n",
       "      <td>-0.1555</td>\n",
       "      <td>0.0445</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.0875</td>\n",
       "      <td>-0.0995</td>\n",
       "      <td>-0.5545</td>\n",
       "      <td>0.132</td>\n",
       "      <td>0.2220</td>\n",
       "      <td>0.2430</td>\n",
       "      <td>0.1830</td>\n",
       "      <td>-0.0040</td>\n",
       "      <td>-0.1085</td>\n",
       "      <td>0.2220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>PDR10</td>\n",
       "      <td>0.0395</td>\n",
       "      <td>-0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0430</td>\n",
       "      <td>-0.162</td>\n",
       "      <td>0.1560</td>\n",
       "      <td>0.2075</td>\n",
       "      <td>0.1125</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0360</td>\n",
       "      <td>0.1130</td>\n",
       "      <td>0.1615</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.2985</td>\n",
       "      <td>-0.0760</td>\n",
       "      <td>-0.2335</td>\n",
       "      <td>0.1015</td>\n",
       "      <td>-0.2010</td>\n",
       "      <td>-0.3375</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 132 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RQTL_name</th>\n",
       "      <th>2_7_a</th>\n",
       "      <th>2_7_b</th>\n",
       "      <th>2_7_c</th>\n",
       "      <th>2_7_d</th>\n",
       "      <th>21_5_c</th>\n",
       "      <th>22_1_d</th>\n",
       "      <th>22_2_d</th>\n",
       "      <th>19_2_c</th>\n",
       "      <th>19_3_c</th>\n",
       "      <th>...</th>\n",
       "      <th>9_7_d</th>\n",
       "      <th>26_2_d</th>\n",
       "      <th>19_1_c</th>\n",
       "      <th>21_3_d</th>\n",
       "      <th>21_1_d</th>\n",
       "      <th>15_6_c</th>\n",
       "      <th>17_1_a</th>\n",
       "      <th>name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YAL069W_1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>1</td>\n",
       "      <td>483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YAL069W_2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>1</td>\n",
       "      <td>484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NAL013C_3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NAL013C</td>\n",
       "      <td>1</td>\n",
       "      <td>3220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NAL013C_4</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NAL013C</td>\n",
       "      <td>1</td>\n",
       "      <td>3223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NAL013C_5</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NAL013C</td>\n",
       "      <td>1</td>\n",
       "      <td>3232</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 128 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RQTL_name</th>\n",
       "      <th>2_7_a</th>\n",
       "      <th>2_7_b</th>\n",
       "      <th>2_7_c</th>\n",
       "      <th>2_7_d</th>\n",
       "      <th>21_5_c</th>\n",
       "      <th>22_1_d</th>\n",
       "      <th>22_2_d</th>\n",
       "      <th>19_2_c</th>\n",
       "      <th>19_3_c</th>\n",
       "      <th>...</th>\n",
       "      <th>9_7_d</th>\n",
       "      <th>26_2_d</th>\n",
       "      <th>19_1_c</th>\n",
       "      <th>21_3_d</th>\n",
       "      <th>21_1_d</th>\n",
       "      <th>15_6_c</th>\n",
       "      <th>17_1_a</th>\n",
       "      <th>name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YAL069W_1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>1</td>\n",
       "      <td>483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YAL069W_2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>1</td>\n",
       "      <td>484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NAL013C_3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NAL013C</td>\n",
       "      <td>1</td>\n",
       "      <td>3220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NAL013C_4</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NAL013C</td>\n",
       "      <td>1</td>\n",
       "      <td>3223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NAL013C_5</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NAL013C</td>\n",
       "      <td>1</td>\n",
       "      <td>3232</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 128 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_df[genotypes_df.columns.tolist()[0:128]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Divide all progeny into groups by their inheritance pattern\n",
    "# for a given genetic marker, and then plot the data clouds\n",
    "# to visually observe if there is any correlation between marker\n",
    "# and RNA expression \n",
    "\n",
    "\n",
    "def expression_by_RNA_and_marker(RNA_name, marker_name):\n",
    "    from_BY, from_RM = [], []\n",
    "    RNA_pos = expression_df[expression_df.IDENTIFIER == RNA_name].index.tolist()[0]\n",
    "    marker_pos = genotypes_df[genotypes_df.RQTL_name == marker_name].index.tolist()[0]\n",
    "    for strain in progeny_names:\n",
    "        expression_value = expression_df[strain].iloc[RNA_pos]\n",
    "        inheritance_pattern = genotypes_df[strain].iloc[marker_pos]\n",
    "        if inheritance_pattern == 0:\n",
    "            from_BY.append(expression_value)\n",
    "        elif inheritance_pattern == 1:\n",
    "            from_RM.append(expression_value)\n",
    "    return from_BY, from_RM\n",
    "\n",
    "\n",
    "def test_linkage(RNA_name, marker_name, eps=1e-5):\n",
    "    from_BY, from_RM = expression_by_RNA_and_marker(RNA_name, marker_name)\n",
    "    statistic, pvalue = stats.mannwhitneyu(x=from_BY, y=from_RM)\n",
    "    return (pvalue <= eps, pvalue)\n",
    "\n",
    "\n",
    "def plot_expression_to_marker_correlation(RNA_name, marker_name):\n",
    "    from_BY, from_RM = expression_by_RNA_and_marker(RNA_name, marker_name)\n",
    "    xlabels = np.append(\n",
    "                    np.full((1, len(from_BY)), 1), \n",
    "                    np.full((1, len(from_RM)), 2))\\\n",
    "                    + np.random.normal(0, 0.01, len(from_BY) + len(from_RM)\n",
    "            )  \n",
    "    ylabels = np.array(from_BY + from_RM)\n",
    "    plt.figure(figsize=(20, 10))\n",
    "    plt.rcParams[\"axes.facecolor\"] = 'white'\n",
    "    plt.title(\"p-value: {}\".format(pvalue))\n",
    "    plt.xlabel(\"class label\")\n",
    "    plt.ylabel(\"expression value\")\n",
    "    plt.scatter(\n",
    "        x=xlabels, y=ylabels,\n",
    "        c=ylabels, cmap=cm.jet\n",
    "    )\n",
    "    plt.savefig(\"./img/\" + RNA_name + \"_to_\" + marker_name + \".png\")\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**TODO**:\n",
    "    1.  Find QTLs for every gene using 5% pvalue threshold\n",
    "    2.  Validate QTLs using permutations test and calculate FDR\n",
    "    3.  Construct and visualize a bipartite graph visualizing \n",
    "        the linkages found during experiments \n",
    "    4.  Also, it's worth calculating the linkages for every\n",
    "        marker gene and plotting the bar chart, placing markers\n",
    "        accordingly to their position in the chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6.659067656776818e-11: Phi174 Control to NKL015C_1698\n"
     ]
    }
   ],
   "source": [
    "# It's quite hard to identify QTLs just by looking at the point clouds.\n",
    "# I guess, the next step is to apply MWU and then plot the pairs with \n",
    "# the highest predicted linkage, just to verify the validity of a program\n",
    "\n",
    "# When and how to apply multiple-testing correction (FDR, q-value)?\n",
    "# Also, it would be interesting to plot a graph of interconnections\n",
    "# between markers and expressed genes \n",
    "\n",
    "np.random.seed(int(time.time()))\n",
    "# Extract subsample\n",
    "test_pairs = list(\n",
    "                zip( \n",
    "                    np.random.choice(expression_df[\"IDENTIFIER\"].tolist(), 1000, replace=True),\n",
    "                    np.random.choice(genotypes_df[\"RQTL_name\"].tolist(), 1000, replace=True)\n",
    "                )\n",
    "            )\n",
    "# Check linkage for all entries of this subsample\n",
    "# and construct a graph of interactions.\n",
    "# Also, the minimal pvalue is also calculated\n",
    "# in this code snippet, it's a temporary thing\n",
    "min_pvalue = 1.\n",
    "argmin_RNA, argmin_marker_name = \"\", \"\"\n",
    "linkage_graph = nx.Graph()\n",
    "for RNA_name, marker_name in test_pairs:\n",
    "    if marker_name.count(\"Control\") > 0:\n",
    "        print(\"WTFWTFWTF\")\n",
    "    result, pvalue = test_linkage(RNA_name, marker_name, eps=0.05)\n",
    "    if result:\n",
    "        if not linkage_graph.has_node(RNA_name):\n",
    "            linkage_graph.add_node(RNA_name, bipartite=0)\n",
    "        if not linkage_graph.has_node(marker_name):\n",
    "            linkage_graph.add_node(marker_name, bipartite=1)\n",
    "        linkage_graph.add_edge(RNA_name, marker_name)\n",
    "        if pvalue < min_pvalue:\n",
    "            min_pvalue = pvalue\n",
    "            argmin_RNA = RNA_name\n",
    "            argmin_marker_name = marker_name\n",
    "            # plot_expression_to_marker_correlation(RNA_name, marker_name)\n",
    "print(\"{}: {} to {}\".format(min_pvalue, argmin_RNA, argmin_marker_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2 PKH2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/vvi/anaconda3/lib/python3.5/site-packages/networkx/drawing/nx_pylab.py:126: MatplotlibDeprecationWarning: pyplot.hold is deprecated.\n    Future behavior will be consistent with the long-time default:\n    plot commands add elements without first clearing the\n    Axes and/or Figure.\n  b = plt.ishold()\n/home/vvi/anaconda3/lib/python3.5/site-packages/networkx/drawing/nx_pylab.py:138: MatplotlibDeprecationWarning: pyplot.hold is deprecated.\n    Future behavior will be consistent with the long-time default:\n    plot commands add elements without first clearing the\n    Axes and/or Figure.\n  plt.hold(b)\n/home/vvi/anaconda3/lib/python3.5/site-packages/matplotlib/__init__.py:917: UserWarning: axes.hold is deprecated. Please remove it from your matplotlibrc and/or style files.\n  warnings.warn(self.msg_depr_set % key)\n/home/vvi/anaconda3/lib/python3.5/site-packages/matplotlib/rcsetup.py:152: UserWarning: axes.hold is deprecated, will be removed in 3.0\n  warnings.warn(\"axes.hold is deprecated, will be removed in 3.0\")\n"
     ]
    }
   ],
   "source": [
    "# Here the linkage graph is analyzed and plotted\n",
    "# It should be visualized as a bipartite graph,\n",
    "# but it is not implemented yet.\n",
    "maxdegree = 0\n",
    "argmaxdegree = \"\"\n",
    "for node in linkage_graph.nodes():\n",
    "    d = linkage_graph.degree(node)\n",
    "    if d > maxdegree:\n",
    "        maxdegree = d\n",
    "        argmaxdegree = str(node)\n",
    "print(maxdegree, argmaxdegree)\n",
    "\n",
    "# top_v, bottom_v = bipartite.sets(linkage_graph)\n",
    "# doesn't work for some strange reasons\n",
    "# I guess, I should either use DiGraph class\n",
    "# or rewrite this as a separate function\n",
    "top_v, bottom_v = [], []\n",
    "for node, data in linkage_graph.nodes(data=True):\n",
    "    if data[\"bipartite\"] == 0:\n",
    "        bottom_v.append(node)\n",
    "    else:\n",
    "        top_v.append(node)\n",
    "# To plot a bipartite graph correctly, the positions\n",
    "# of the vertices must be written down explicitly\n",
    "pos = dict()\n",
    "pos.update((n, (1, 2*i)) for i, n in enumerate(top_v))\n",
    "pos.update((n, (2, 2*i)) for i, n in enumerate(bottom_v))\n",
    "\n",
    "# Maybe, (20, 20) is too small: the labels overlap significantly\n",
    "plt.figure(figsize=(20, 30))\n",
    "nx.draw(\n",
    "    linkage_graph,\n",
    "    with_labels=True,\n",
    "    node_size=50,\n",
    "    edge_width=3.0,\n",
    "    pos=pos,\n",
    "    node_color=list(linkage_graph.degree().values()),\n",
    "    edge_color='b',\n",
    "    cmap=plt.cm.Blues,\n",
    "    alpha=0.5\n",
    ")\n",
    "plt.savefig(\"./img/graph.png\")\n",
    "plt.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
