{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "import collections\n",
    "import itertools\n",
    "from importlib import reload\n",
    "import pickle\n",
    "\n",
    "# data analysis tools\n",
    "import random\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "\n",
    "# os.chdir(\"/home/ivanov_vv/eQTL_analysis/\")\n",
    "\n",
    "import networks\n",
    "import qtls\n",
    "import util\n",
    "import ontologies\n",
    "\n",
    "# 112 segregants genotyped by inherited marker variants\n",
    "full_genotypes_df = pd.read_table(\"./data/genotypes/processed_genotypes.csv\")\n",
    "\n",
    "''' Where possible, gene names were converted from systematic to standard notation '''\n",
    "# mRNA expression and genotypes of strains the data is available for\n",
    "eQTLs_expression_df = pd.read_table(\"./data/eQTLs/averaged_expression.csv\")\n",
    "eQTLs_genotypes_df = pd.read_table(\"./data/eQTLs/processed_genotypes.csv\")\n",
    "eQTLs_genotypes_df.drop_duplicates(\"SNP\", inplace=True)\n",
    "# Protein expression and genotypes of strains the data is available for \n",
    "pQTLs_expression_df = pd.read_table(\"./data/pQTLs/averaged_expression.csv\")\n",
    "pQTLs_genotypes_df = pd.read_table(\"./data/pQTLs/processed_genotypes.csv\")\n",
    "pQTLs_genotypes_df.drop_duplicates(\"SNP\", inplace=True)\n",
    "\n",
    "# pandas DataFrame â€”> numpy matrix = 2x speedup\n",
    "pQTLs_expression_mx = pQTLs_expression_df.as_matrix(columns=pQTLs_expression_df.columns[1:])\n",
    "pQTLs_genotypes_mx = pQTLs_genotypes_df.as_matrix(columns=pQTLs_genotypes_df.columns[1:])\n",
    "\n",
    "# QTLs estimated with MatrixEQTL package for R\n",
    "eQTLs_df = pd.read_table(\"./data/eQTLs/results.csv\")\n",
    "eQTLs_df = eQTLs_df[eQTLs_df[\"q.value\"] <= 0.05]\n",
    "pQTLs_df = pd.read_table(\"./data/pQTLs/results.csv\")\n",
    "pQTLs_df = pQTLs_df[pQTLs_df[\"q.value\"] <= 0.05]\n",
    "\n",
    "# Physical and genetic interactions taken from TheBioGrid project\n",
    "full_interactome_df = pd.read_table(\"./data/interactions/yeast_interactome.csv\")\n",
    "# full_interactome_df = full_interactome_df[full_interactome_df['Publication year'] <= 2011]\n",
    "\n",
    "'''\n",
    "Move these to their project part, there is no need for them anywhere else\n",
    "'''\n",
    "genetic_interaction_categories = [\n",
    "    \"additive genetic interaction defined by inequality\",\n",
    "    \"suppressive genetic interaction defined by inequality\",\n",
    "    \"synthetic genetic interaction defined by inequality\"\n",
    "]\n",
    "physical_interaction_categories = [\n",
    "    \"association\",\n",
    "    \"colocalization\",   \n",
    "    \"direct interaction\",\n",
    "    \"physical association\"\n",
    "]\n",
    "\n",
    "interaction_categories = genetic_interaction_categories + physical_interaction_categories\n",
    "interactome_df = full_interactome_df.copy()\n",
    "interactome_df.set_index('Interaction Type', inplace=True)\n",
    "genetic_interactions_df = interactome_df[\n",
    "    interactome_df.index.isin(genetic_interaction_categories)\n",
    "]\n",
    "physical_interactions_df = interactome_df[\n",
    "    interactome_df.index.isin(physical_interaction_categories)\n",
    "]\n",
    "\n",
    "def extract_interactions(interactions_df):\n",
    "    ''' Helper function to retrieve the edges of the graph of interactions '''\n",
    "    return interactions_df[[\"Interactor A\", \"Interactor B\"]].values\n",
    "\n",
    "interaction_graphs_dict = {}\n",
    "for category_name in interaction_categories:\n",
    "    interacting_genes_df = interactome_df[interactome_df.index == category_name]\n",
    "    interaction_graphs_dict[category_name] = networks.graph_from_edges(\n",
    "        extract_interactions(interacting_genes_df),\n",
    "    )\n",
    "    \n",
    "interaction_graphs_dict.update(dict(zip([\"all\", \"genetic\", \"physical\"],\n",
    "    itertools.starmap(networks.graph_from_edges, \n",
    "        [(extract_interactions(interactome_df), False, False),\n",
    "         (extract_interactions(genetic_interactions_df), False, False), \n",
    "         (extract_interactions(physical_interactions_df), False, False)]\n",
    "    )))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with open(\"./results/kegg_pathways/modules_dict.pkl\", \"wb+\") as pickle_file:\n",
    "#     pickle.dump(kegg_pathways_dict, pickle_file)\n",
    "with open(\"./results/kegg_pathways/modules_dict.pkl\", \"rb+\") as pickle_file:\n",
    "    kegg_pathways_dict = pickle.load(pickle_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(qtls)\n",
    "qtls.process_ontologies(\n",
    "    database_name=\"kegg_pathways\",\n",
    "    expression_dfs=[eQTLs_expression_df, pQTLs_expression_df],\n",
    "    qtl_dfs=[eQTLs_df, pQTLs_df],\n",
    "    qtl_types=[\"eQTLs\", \"pQTLs\"],\n",
    "    interactome_graph=interaction_graphs_dict[\"physical\"],\n",
    "    module_gene=kegg_pathways_dict,\n",
    "    q_thresholds=np.logspace(-5, -2, 10)\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
